django/db/models/query_utils.py

```diff
@@ -340,38 +340,37 @@ def _unregister_instance_lookup(self, lookup, lookup_name=None):
     _unregister_class_lookup = classmethod(_unregister_class_lookup)
 
 
-def select_related_descend(field, restricted, requested, select_mask, reverse=False):
+def select_related_descend(field, restricted, requested, select_mask):
     """
-    Return True if this field should be used to descend deeper for
-    select_related() purposes. Used by both the query construction code
-    (compiler.get_related_selections()) and the model instance creation code
-    (compiler.klass_info).
+    Return wether `field` should be used to descend deeper for
+    `select_related()` purposes.
 
     Arguments:
-     * field - the field to be checked
-     * restricted - a boolean field, indicating if the field list has been
-       manually restricted using a requested clause)
-     * requested - The select_related() dictionary.
-     * select_mask - the dictionary of selected fields.
-     * reverse - boolean, True if we are checking a reverse select related
+     * `field` - the field to be checked. Can be either a `Field` or
+       `ForeignObjectRel` instance.
+     * `restricted` - a boolean field, indicating if the field list has been
+       manually restricted using a select_related() clause)
+     * `requested` - the select_related() dictionary.
+     * `select_mask`
     """
+    # Only relationships can be descended.
     if not field.remote_field:
         return False
-    if field.remote_field.parent_link and not reverse:
+    # Forward MTI parent links should not be explicitly descended as they are
+    # always JOIN'ed against unless excluded by `select_mask`.
+    if getattr(field.remote_field, "parent_link", False):
         return False
-    if restricted:
-        if reverse and field.related_query_name() not in requested:
-            return False
-        if not reverse and field.name not in requested:
-            return False
-    if not restricted and field.null:
+    # When `select_related()` is used without a *requested mask all
+    # relationships are descended unless they are nullable.
+    if not restricted:
+        return not field.null
```

Comment:
[Per the docs](https://docs.djangoproject.com/en/5.0/ref/models/querysets/#django.db.models.query.QuerySet.select_related)

> There may be some situations where you wish to call select_related() with a lot of related objects, or where you don’t know all of the relations. In these cases it is possible to call select_related() with no arguments. This will follow all **non-null foreign keys** it can find - **nullable foreign keys must be specified**.